internal/pkg/api/handleEnroll.go

```diff
@@ -252,6 +253,85 @@ func (et *EnrollerT) _enroll(
 				Msg("Error when trying to delete old agent with enrollment id")
 			return nil, err
 		}
+		// deleted, so clear the ID so code below knows it needs to be created
+		agent.Id = ""
+	}
+
+	var agentID string
+	if req.Id != nil && *req.Id != "" {
+		agentID = *req.Id
+
+		// check if the agent with this ID already exists
+		var err error
+		agent, err = et._checkAgent(ctx, zlog, agentID)
+		if err != nil {
+			return nil, err
+		}
+
+		if agent.Id != "" {
+			// confirm that this agent has a set replace token
+			// one is required or replacement of this already enrolled and active
+			// agent is not allowed
+			if agent.ReplaceToken == "" {
+				zlog.Warn().
+					Str("AgentId", agent.Id).
+					Msg("Existing agent with same ID already enrolled without a replace token set")
+				return nil, ErrAgentNotReplaceable
+			}
+			if req.ReplaceToken == nil || *req.ReplaceToken == "" {
+				zlog.Warn().
+					Str("AgentId", agent.Id).
+					Msg("Existing agent with same ID already enrolled; no replace token given during enrollment")
+				return nil, ErrAgentNotReplaceable
+			}
+			same, err := compareHashAndToken(zlog.With().Str("AgentID", agent.Id).Logger(), agent.ReplaceToken, *req.ReplaceToken, et.cfg.PDKDF2)
+			if err != nil {
+				// issue with hash comparison; reason already logged
+				return nil, ErrAgentNotReplaceable
+			}
+			if !same {
+				// not the same, cannot replace
+				// provides no real reason as that would expose too much information
+				zlog.Debug().
+					Str("AgentId", agent.Id).
+					Msg("Existing agent with same ID already enrolled; replace token didn't match")
+				return nil, ErrAgentNotReplaceable
+			}
+
+			// confirm that its on the same policy
+			// it is not supported to have it the same ID enroll into different policies
+			if agent.PolicyID != policyID {
+				zlog.Warn().
+					Str("AgentId", agent.Id).
+					Str("PolicyId", policyID).
+					Str("CurrentPolicyId", agent.PolicyID).
+					Msg("Existing agent with same ID already enrolled into another policy")
+				return nil, ErrAgentNotReplaceable
+			}
+
+			// invalidate the previous api key
+			// this has to be done because it's not possible to get the previous token
+			// so the other is invalidated and a new one is generated
+			zlog.Debug().
+				Str("AgentId", agent.Id).
+				Str("APIKeyID", agent.AccessAPIKeyID).
+				Msg("Invalidate old api key with same id")
+			err = invalidateAPIKey(ctx, zlog, et.bulker, agent.AccessAPIKeyID)
```

Comment:
Does any of the state in the .fleet-agents document need to be reset? For example the new agent is not going to have the same persisted action state or .fleet-action sequence number, and will not yet have a copy of the policy. 

If we do any state resets there, it has to happen after the API key confirmed invalidated and not just when it is requested to avoid race conditons.

Reply:
The `.fleet-action` sequence number I do not reset, as I don't think we want this Elastic Agent to replay all actions of the previous Elastic Agent. If this is used outside of the container use-case, you would not expect it to roll through performing an upgrade action because it replaced a previous Elastic Agent.

I took this into account and didn't reset the sequence number for that exact reason. The policy and revision is not tied to the sequence number in anyway. It is actually stored directly on the `.fleet-agents` document, which is reset by this code here - https://github.com/elastic/fleet-server/pull/4290/files#diff-b3d2d3a49ef214fba08e3aa10772162cbb65c378e24ef6c4a0c5df65b051f71aR378